### PR TITLE
[fix][broker] Fix messages could not expire due to incorrect client clock

### LINKED ISSUE #21347: [DISK Dos Attack Warning] should not use message publish time in checkMessageExpiry and AppendBrokerTimesampMetadataInterceptor should be config to brokerEntryMetadataInterceptors
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Version

all pulsar version

### Minimal reproduce step

1、brokerEntryMetadataInterceptors config not contain AppendBrokerTimesampMetadataInterceptor 
2、create a topic and set retention and ttl to 5 min
3、pulsar client server change system time to some years later
4、publish message and start a pulsar consumer
5、stop  pulsar consumer and continue publish message 
6、message can not expire when ledger close and time reach

### What did you expect to see?

NA

### What did you see instead?

PersistentMessageExpiryMonitor 
expireMessages(int messageTTLInSeconds)  -->call Commands/MessageImpl.getEntryTimestamp(entry.getDataBuffer());

```
// get broker timestamp first if BrokerEntryMetadata is enabled with AppendBrokerTimestampMetadataInterceptor
BrokerEntryMetadata brokerEntryMetadata =Commands.parseBrokerEntryMetadataIfExist(headersAndPayloadWithBrokerEntryMetadata);
if (brokerEntryMetadata != null && brokerEntryMetadata.hasBrokerTimestamp()) {
    return brokerEntryMetadata.getBrokerTimestamp();
 }
// otherwise get the publish_time
return Commands.parseMessageMetadata(headersAndPayloadWithBrokerEntryMetadata).getPublishTime();
```

when brokerEntryMetadata.hasBrokerTimestamp() == false , pulsar server will use mesage publishTime
this will introduce vulnerability attack , server can not trust any client message data

Suggestions：
add AppendBrokerTimesampMetadataInterceptor to brokerEntryMetadataInterceptors

### Anything else?

Others:
in transaction scenario , even when we config brokerEntryMetadataInterceptors=org.apache.pulsar.common.intercept.AppendIndexMetadataInterceptor, we can not get offset , i am not sure if brokerEntryMetadataInterceptors works in transaction produce

if brokerEntryMetadataInterceptors not work 
maybe we need verify client publish time when server receive mesage .

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by 315157973**
Fixes #21347

### Motivation
When the client machine's clock is incorrect (eg: set to 1 year later) and the Broker does not set the AppendBrokerTimestampMetadataInterceptor, the Ledger will not be cleaned up.
Because if the Broker's timestamp is not set, the expiration check will be based on the client's publish time.

### Modifications
Always set the Broker's timestamp

### Verifying this change

- [x] Make sure that the change passes the CI checks.

### Documentation
- [x] `doc-not-needed` 

### Matching PR in forked repository

PR in forked repository: https://github.com/315157973/pulsar/pull/12

### CODE REVIEW

**Review by Technoboy- — APPROVED** [review]

**Review by lhotari — CHANGES_REQUESTED** [review]
Please also fix another issue with the reuse of the `brokerEntryMetadata`.
Replace `BrokerEntryMetadata brokerEntryMetadata = BROKER_ENTRY_METADATA.get();` with `BrokerEntryMetadata brokerEntryMetadata = BROKER_ENTRY_METADATA.get().clear();` to ensure that the brokerTimestamp doesn't get reused from the previous time the instance was used. Otherwise the timestamp would never get updated.

**Comment by lhotari** [comment]
Please also update the comment in `org.apache.pulsar.common.protocol.Commands#getEntryTimestamp` . I guess that now the BrokerEntryMetadata with the timestamp will always be added?

File: pulsar-common/src/main/java/org/apache/pulsar/common/protocol/Commands.java [hunk 1]
```diff
@@ -1697,13 +1697,16 @@ public static ByteBuf addBrokerEntryMetadata(ByteBuf headerAndPayload,
         //   | BROKER_ENTRY_METADATA_MAGIC_NUMBER | BROKER_ENTRY_METADATA_SIZE |         BROKER_ENTRY_METADATA         |
         //   |         2 bytes                    |       4 bytes              |    BROKER_ENTRY_METADATA_SIZE bytes   |
 
-        BrokerEntryMetadata brokerEntryMetadata = BROKER_ENTRY_METADATA.get();
+        BrokerEntryMetadata brokerEntryMetadata = BROKER_ENTRY_METADATA.get().clearBrokerTimestamp();
```
Comment (lhotari) [inline pulsar-common/src/main/java/org/apache/pulsar/common/protocol/Commands.java:1700]: Please call `.clear()` to reset all state.

**Review by lhotari — CHANGES_REQUESTED** [review]

File: pulsar-common/src/main/java/org/apache/pulsar/common/protocol/Commands.java [hunk 2]
```diff
@@ -1697,13 +1697,16 @@ public static ByteBuf addBrokerEntryMetadata(ByteBuf headerAndPayload,
...
         for (BrokerEntryMetadataInterceptor interceptor : brokerInterceptors) {
             interceptor.intercept(brokerEntryMetadata);
             if (numberOfMessages >= 0) {
                 interceptor.interceptWithNumberOfMessages(brokerEntryMetadata, numberOfMessages);
             }
         }
+        if (!brokerEntryMetadata.hasBrokerTimestamp()) {
+            brokerEntryMetadata.setBrokerTimestamp(System.currentTimeMillis());
+        }
```
Comment (codelipenghui) [inline pulsar-common/src/main/java/org/apache/pulsar/common/protocol/Commands.java:1709]: The broker timestamp will be set in line:1702 if the `AppendBrokerTimestampMetadataInterceptor` is enabled. Is this PR wants to set the broker timestamp without `AppendBrokerTimestampMetadataInterceptor` enabled?

Comment (315157973) [reply pulsar-common/src/main/java/org/apache/pulsar/common/protocol/Commands.java:1709]: Unless we can ensure that the `AppendBrokerTimestampMetadataInterceptor` cannot be disabled, this bug may always occur

**Review by lhotari — APPROVED** [review]

File: pulsar-common/src/main/java/org/apache/pulsar/common/protocol/Commands.java (hunk 2 above)
Comment (codelipenghui) [reply pulsar-common/src/main/java/org/apache/pulsar/common/protocol/Commands.java:1709]: Ok, I think we can either make the AppendBrokerTimestampMetadataInterceptor enabled by default or remove the AppendBrokerTimestampMetadataInterceptor to always apply the broker timestamp.

Without this change, the broker timestamp will be enabled without AppendBrokerTimestampMetadataInterceptor. Essentially, we delete the functionality of the AppendBrokerTimestampMetadataInterceptor.

Comment (315157973) [reply pulsar-common/src/main/java/org/apache/pulsar/common/protocol/Commands.java:1709]: I opened a discuss in dev thread https://lists.apache.org/thread/0vtjt436xvlg0fr4ogchkjdvjcp56mvz

**Comment by 315157973** [comment]
@lhotari @codelipenghui  Can this PR be merged?

**Comment by 315157973** [comment]
As https://github.com/apache/pulsar/pull/21940 is merged, close this PR

**Pull request closed without merging**